Getting started tutorials

cuihtlauac opened the pull request:
The Getting Started series of tutorial include the basic prerequisites
needed to begin learning OCaml
1. Install
1. A Tour of OCaml
1. How to Write an OCaml Program

Anybody learning OCaml, student, self-learner, or seasoned developer should walk through these tutorials to get the basic concepts and structure them. This is intended to create a breadth-first learning experience. No deep dive into anything, just running above the most important topics.

Follow-up should be **lists** (for those unfamiliar with them) or “Pro Setup” for experienced developers.

Commit:
Replace 1st hour by Tour of OCaml

This is the new version of the first tutorial that should be read
right after installing OCaml. It focuses on interaction with OCaml
using utop. Languages features are showcased, nothing is presented
in detail.

Commit:
How to Write an OCaml Program

This tutorial is a follow-up from “A Tour of OCaml”, it contains some of the materials contained in “First hour/day with OCaml” but also significant changes. In contrast to the previous of this series, this tutorial is centered on the “working with files“ story: writing .ml and .mli files, modules, packages, compilation and dune.

However, as in the previous document, this a tour, not a detailed guide. This is not sufficient to be a proficient OCaml developer, but it is absolutely required to have an understanding of the materials presented here to start building a learning path into OCaml.

Commit:
Fix build

christinerose reviewed (commented):
Comment about wording of "big five game"

Commit:
Review from @christinerose

christinerose committed:
changing a word to British spelling

christinerose committed:
Edits to intro paragraphs

christinerose committed:
Edits through line 60

christinerose committed:
Edits to li 81

christinerose committed:
edits to line 85

christinerose committed:
Typos in li 85

christinerose committed:
Edits li 87-94

christinerose committed:
Editing through li 134

christinerose committed:
Edits through li 170

christinerose committed:
Edits through 258

christinerose committed:
fix misunderstanding in li 222

christinerose committed:
fix grammatical error in 243

christinerose committed:
Editing through the end

christinerose committed:
Fix opam formatting

christinerose committed:
typo and unneeded word in 260

christinerose committed:
correct sentence structure/grammar in 273

christinerose committed:
missing word in li 300

christinerose reviewed (commented):
Apologies for all the noise and multiple commits! As I went through line editing, there were things I needed to comment on for clarity, so I'd commit to comment before I forgot. 

Also, I made of a note of where I stopped testing the code because I was getting so many errors. Either I'm doing something wrong (which might indicate parts need clarification) or there's something off in the code. I'll go through it will @cuihtlauac soon.

christinerose committed:
Apply suggestions from code review

christinerose reviewed (commented):
Stopped at 215 for the day. Will pick up tomorrow!

christinerose reviewed (commented):
About possible confusion around `#`, although readers are probably smart enough to figure it out. 🙂

christinerose reviewed (commented):
I'm still getting too many errors, even after we updated my system for ARM64. Will work with @cuihtlauac on this.

christinerose committed:
Clarifications in code instructions

christinerose committed:
fix output to remove extra Dune noise

Commit:
More review from @christinerose

cuihtlauac committed:
Apply suggestions from @sabine and @christinerose

Co-authored-by: sabine <sabine@users.noreply.github.com>
Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

Commit:
More edits

Commit:
Handle some FIXME

Commit:
Yet another note

Commit:
Fix doc id

Commit:
Fix links

Commit:
Fix category

Commit:
Various edits

sabine reviewed (commented):
File names need to be such that the two new tutorials appear after the two existing ones ("Getting Started", "OCaml on Windows").

Commit:
Rename file, restore order

Commit:
Add learning goals

christinerose reviewed (commented):
This looks good other than changing the placeholder links to the actual links :)

sabine committed:
rebase rename

sabine committed:
tutorial blocks point to new tutorials

sabine committed:
fmt

sabine committed:
titles and minor intro changes

sabine committed:
update install.eml

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+# if 0 = 1 then "test is true" else "test is false";;
+- : string = "test is false"
+```
+In OCaml, `if ... then ... else ...` and the ternary conditional operator are the same thing.
````
I think the if expression behaviour might be very unfamiliar/subtle to some. I suggest that it comes after the introduction to assigning variables (`let x = 50`) and include an example of assigning the result of an if expression to one.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -0,0 +1,768 @@
+
+_FIXME: Shall we move this to another tutorial? Which one?_
+
+Operators are functions. The function underlying behind an operator is refered by surrounding the operator symbol with parenthesis. Here are the addition, string concatenation and equality functions:
```
This is interesting, but I’m guessing it’s on the low use scale. Probably better served in standard docs.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+
+# range 2 11;;
+- : int list = [2; 3; 4; 5; 6; 7; 8; 9; 10; 11]
+```
````
Likely to throw the uninitiated for a doozy here. Super important for the functional programming side of OCaml, but it might need more priming. Also, the  `::` operator hadn’t been introduced yet. It looks like it adds a single value to the beginning of a list, but I had to do some extra legwork to figure that out.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+    | 1 -> 1
+    | n -> n * fact (n - 1);;
+val fact : int -> int = <fun>
+```
````
I think it would be useful to provide a little bit more explanation on what is happening here. Also worth considering using  a simpler example that focuses on the core feature, and then  following it up with this example that includes recursion. 

For the catch all case in this example, would it be useful to use an alternative variable name? I assume it is shadowing the function argument, but it might be confusing seeing `n` everywhere. Might be an OCaml convention, but think it could be subtly confusing.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+
+# sum [1; 4; 3; 2; 5]/L
+- : int = 15
+```
````
I feel like this section should come before “Pattern Matching”. The pattern matching could be introduced straight afterwards, with the sum example being used as an example. It’s nice when a lesson builds on top of a prior lesson.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+
+# let u = range 0 10000 in List.map square u = map square u;;
+- : bool = true
+```
````
The final statement in this example is too complicated. I couldn’t quite figure out the two separate `=` operators on the same line. The `in` operator is also new to me. Wondering if this could be simplified to just use the square function.

```ocaml
let square x = x * x;;

let rec map f u =
  match u with
  | [] -> []
  | x :: u -> f x :: map f u;;

let squared = map square [1; 2; 3; 4;];;
```

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+    match p with
+    | (_, y) -> y;;
+val snd : 'a * 'b -> 'b = <fun>
+```
````
I personally prefer when an example also shows how to use it. It makes it much easier for me to understand whats going on as I get to also reverse engineer it in my mind. I also prefer to avoid abbreviations, especially when writing content for newcomers as it helps avoid unnecessary assumptions/confusion.

```ocaml
let second_element p =
    match p with
    | (_, y) -> y;;

let fruits = ("apple", "pear");;

let x = second_element fruits;;
(* val x : string = "pear"; *)
```

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+| Current
+| Range of int * int;;
+type page_range = All | Current | Range of int * int
+```
````
I feel like it would be helpful to show how to create instances of these variant types.

Looks like an incredibly powerful feature. Love it.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -0,0 +1,768 @@
+     surname = "Huet";
+     age = 40
+  };;
+val frank : person = {first_name = "Gérard"; surname = "Huet"; age = 40}
```
I heard of this before, but I think it is worth calling out that OCaml automatically associates a type definition with record instances that match a type. Really cool feature, but would be especially confusing for all the TypeScript peeps out there.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac replied:
Addressed in: c79254cdb87cb9f9f5dbce8d32e669a7e73bb924

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -0,0 +1,768 @@
+val is_teenager : person -> bool = <fun>
+
+# is_teenager frank;;
+- : bool = false
```
Ooooh, this is cool. So would this match any record type that had an `age` property along with any other number of properties?

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac replied:
Addressed in: c79254cdb87cb9f9f5dbce8d32e669a7e73bb924

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+
+# id_42_res 0;;
+- : (int, string) result = Error "Sorry"
+```
````
Wonder if it is worth showing how you can use pattern matching against the `result` type to emulate the try..catch example that was previously show?

```ocaml
let n = match id_42_res 42 with
  | Ok n -> n
  | Error _ -> 0;;
```

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+```
+
+Here the first side-effect displays `hello `, the second `world!`. The
+returned value comes from the second function call.
````
Are there any more solid real world examples for why this is useful? The given example appears a bit arbitrary to me right now.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,768 @@
+
+# Option.map (fun x -> x * x) (Some 8);;
+- : int option = Some 64
+```
````
I am wondering if people who are unfamiliar with curried functions might get confused on some of this stuff. I think you do cover it a little in previous sections, but I am starting to wonder if it needs it’s own expanded section given it’s likely such a fundamental part of using the language effectively? 

For this example code it might be worth introducing variables to make this clearer?

```ocaml
let opt_multiply = Option.map (fun x -> x * x);;

opt_multiply None;;

opt_multiply (Some 8);;
```

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
Reading the types output by the REPL can be a trip up, especially for new comers. I think it’s important that developers are able to understand the type definitions though. Perhaps a little section on this, or have it interlaced with the other sections.

From: https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

Commit:
Some fixes

Address the easiest issues highlighted in:

https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

Thanks to the anonymous reviewer.

Commit:
Expand recursive functions text

Commit:
Add text on record type checking

Commit:
Clarify text

Commit:
Address remaining sockthedev.notion.site feedback

See:
https://sockthedev.notion.site/A-Tour-of-OCaml-Feedback-e5f3e62f445246cd8725c30b850cdb45

Commit:
Add missing text

Commit:
Add text on equality

Commit:
Fix module interface text

christinerose committed:
slight syntax change

sabine committed:
add 'Installing OCaml' tutorial

sabine committed:
show list

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>
Co-authored-by: Dmitrii Kosarev <Kakadu@pm.me>
Co-authored-by: Kakadu <Kakadu@users.noreply.github.com>

sabine committed:
remove 'up and running'

Commit:
Add text on switches

chshersh reviewed (approved):
Really great stuff!

I added a few suggestions from my teaching experinece but feel free to ignore if you have different preferences 😌 

P.S. I haven't read previous review comments, so apologies for suggesting something that was done initially and was changed later due to review feedback 🙏🏻

Commit:
Changes requested by Gregg Reynolds

Commit:
Add epigraph

Commit:
Add epigraph, cont'd

cuihtlauac commented on data/tutorials/getting-started/gs_02_fst_program.md:
````diff
@@ -0,0 +1,343 @@
+
+In addition to that, modules allows organising projects. Concerns are separated into isolated modules. This is outlined in the next section. Before creating a module, here is how to use a definition from a module in our project.
+```shell
+$ echo 'let () = Printf.printf "%s!\\n" "Hello, World!"' > bin/main.ml
````
You are correct. I had planned to remove those command lines. But forgot to do. Yes, copying and pasting is better.

cuihtlauac replied:
Fixed in 1820822e

Commit:
No longer edit .ml file from command line

cuihtlauac commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,870 @@
+- : string = "hi friend"
+```
+
+### Type Variables and Higher-Order Functions
````
I am split on this.

On the one hand, I fully agree with you these are different topics, and there is much to learn by gradually exposing things as you suggest. This is also the way I used to teach this. It works and does good.

On the other hand, I also want this tutorial to be as short as possible. This section aims to show `map` in OCaml because, nowadays, this is something readers will probably already know. This is intended to be a top-down thing, showing `map`, how it's used, and mentioning how it relates to polymorphism and higher-order without explaining much about any.

What do you think about that?

Commit:
Fix epigraphs

cuihtlauac committed:
Apply suggestions

From @tuohy and @christinerose

Co-authored-by: tuohy <g.tuohy@outlook.com>
Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

cuihtlauac committed:
Apply suggestions from code review

Co-authored-by: tuohy <g.tuohy@outlook.com>

cuihtlauac committed:
Update data/tutorials/getting-started/gs_02_fst_program.md

Co-authored-by: tuohy <g.tuohy@outlook.com>

Commit:
Tiny rewording

christinerose committed:
slight rewording for clarification

Commit:
Add epigram

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -0,0 +1,874 @@
+
+> Franz avait lu dans France Soir qu’un Américain avait mis neuf minutes et quarante-cinq secondes pour visiter le musée du Louvre. Ils décidèrent de faire mieux…
+
+&mdash; Jean-Luc Godard, _Bande à part_
```
```suggestion

```

Thibaut suggested we remove this.

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -0,0 +1,874 @@
+
+There are syntactic restrictions on possible operators symbols. They will not be
+detailed here.
+
```
```suggestion

```

Thibaut suggests we move this into a tutorial of its own, or at least into another one.

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
````diff
@@ -0,0 +1,874 @@
+# a.(4);;
+- : int = 0
+```
+
````
```suggestion

```

Thibaut suggested we move this to another tutorial and put a link here, something like:

When you're ready, explore other topics like [dealing with errors](link), [working in a mutable state](link), and using [operators as functions](link) 

(as an example, please feel free to revise!)

christinerose reviewed (commented):
Some suggestions from my meeting with Thibaut yesterday (as I understood them, so please forgive any misunderstandings!). We're almost there! Fingers crossed we can get these first few docs out this week!

christinerose committed:
More concise global rewrite

christinerose committed:
Create gs_04_arm_fix.md

christinerose committed:
Create gs_05_opam_switch.md

christinerose committed:
changed link to new install tutorial

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -0,0 +1,874 @@
+- Raise and catch an exception
+- Return an error-as-data result, process it using pattern matching
+- Declare and update a mutable basic mutable values: references and arrays
+- Call functions defined in modules of the OCaml standar library
```
```suggestion
Before proceeding with this tutorial, please ensure you've installed OCaml and set up the environment, as described on the [install OCaml](/docs/installing-ocaml) page. After we take an introductory tour of OCaml's language features, we'll proceed to create our first OCaml project in the [How to Write an OCaml Program](/docs/how-to-write-an-ocaml-program) tutorial.

This is a level zero tutorial. No OCaml or functional programming knowledge is required. Although, it is assumed the reader has some basic software development knowledge. 

This document will cover how to use the REPL UTop to evaluate OCaml expressions interactively, understand the output, how to use pattern matching, call functions in OCaml standard library modules, and more. It also introduces you to how lists, values, functions, integers, floats, basic operators, references, and arrays work in OCaml.
```

While it's often beneficial to put what we'll be learning at the front of a tutorial/lesson, Thibaut felt this list was too long, and it might discourage/overwhelm the reader. This is my suggestion for edits.

Also, Thibaut suggested we put the "prerequisites" at the top and the "next up" at the bottom as parts of the structure rather than the text. For example, the bottom of each page will have a link to what's next on the right-hand side and a link to the one before on the left.

>should we put in a link to where they can learn code? I don't think it's necessary, perhaps. They can Google for sure, unless we'll have a basic intro somewhere we can link to in the future.

christinerose committed:
added links to ARM64 fix & opam switches

christinerose committed:
fix misspelling

christinerose committed:
fix syntax & verb tense

christinerose committed:
added link to Installing OCaml

christinerose committed:
removed unnecessary link

sabine committed:
remove date

sabine commented on data/tutorials/getting-started/gs_04_arm_fix.md:
```diff
@@ -0,0 +1,69 @@
+title: ARM64 Fix
+description: |
+  This page will walk you through the workaround for ARM64 processors on newer Macs.
+category: "getting-started"
```
```suggestion
category: "Troubleshooting"
```

I would recommend to rename files and have a troubleshooting section because this frames that you want to look at this only when there's trouble.

sabine commented on data/tutorials/getting-started/gs_04_arm_fix.md:
```diff
@@ -0,0 +1,69 @@
+---
+id: arm64-fix
+title: ARM64 Fix
```
May need a better title, this title is not specific enough for people to know if this is for them. What are people seeing on their screen and looking a fix for?

christinerose replied:
That's a really good question. I believe we just kept getting errors because Homebrew installed something in the wrong place. I'm not sure how to word that for a title. Perhaps @cuihtlauac has an idea?

UPDATE: Fixed.

christinerose committed:
Create gs_06_editor_setup

Commit:
Team update of install

christinerose committed:
opam formatting

Commit:
Address easy remarks from @thibault

Commit:
Minor updates

christinerose committed:
Update gs_01_a_tour_of_ocaml.md

christinerose committed:
formatting opam

christinerose committed:
formatting Dune

christinerose committed:
change title and category

Commit:
Update introduction

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -14,9 +14,9 @@ Let's walk through the basics of OCaml by trying out different elements in an in
 Before proceeding with this tutorial, please ensure you've installed OCaml and set up the environment, as described on the [install OCaml](/docs/installing-ocaml) page. After we take an introductory tour of OCaml's language features, we'll proceed to create our first OCaml project in the [How to Write an OCaml Program](/docs/how-to-write-an-ocaml-program) tutorial.
 
-This is a level zero tutorial. No OCaml or functional programming knowledge is required. As mentionned above, only having OCaml installed is needed. It is assumed the reader has some basic software development knowledge. This tutorial is probably not adapted to learn programming.
+This is a level zero tutorial. No OCaml or functional programming knowledge is required. Only having OCaml installed is needed. It is assumed the reader has some basic software development knowledge. This tutorial is probably not adapted to learn programming.
```
```suggestion
To participate in this level zero tutorial, you need to have OCaml installed. No OCaml or functional programming knowledge is required; however, it is assumed the reader has some basic software development knowledge. This tutorial is probably not adapted to learn programming.
```

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -51,21 +51,21 @@ utop #
 Press `Ctrl-D` (end of file) or enter `#quit;;` to exit `utop`.
 
-The OCaml code examples in this tutorial show the input needed at the hash prompt `#`, as displayed by `utop`, ending with double semicolumn `;;`. The `;;` let OCaml know the expression is complete. Code sample lines beginning with dollar sign `$`, like above, should be entered in an terminal.
+The hash prompt `#` displayed by UTop means it is waiting for input. You can start writing your code after the prompt. To evaluate it, add a double semicolon `;;` to signal the end of the expression and press `Enter`.
```
```suggestion
UTop displays a hash prompt `#`, similar to the `$` in the CLI. This `#` means it is waiting for input, so can start writing your code after the prompt. To evaluate it, add a double semicolon `;;` to signal the end of the expression and press `Enter`.
```

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -51,21 +51,21 @@ utop #
+If you need to amend the code before hitting `Enter`, you can use your keyboard's right and left arrows to move inside the text. The up and down arrows allows navigation through previously evaluated expressions. Typing `Enter` without a double semicolon `;;` will create a new line: you can write multiple-line expressions this way.
 
-If you need to amend the code before hitting enter, you can use your keyboard's right and left arrows to move inside the text. The up and down arrows allows navitation through previously typed commands, if you have multiple lines of code.
+Commands beginning with a dash character `#` such as `#quit` or `#help` are not evaluated by OCaml: they are interpreted as commands by UTop.
```
Do you mean "dash" here or "hash" ?

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -51,21 +51,21 @@ utop #
+Commands beginning with a dash character `#` such as `#quit` or `#help` are not evaluated by OCaml: they are interpreted as commands by UTop.
 
-Remember: commands beginning with a dash character `#` such as `#quit` or `#help` are not evaluated by OCaml, they are interpreter by Utop.
+You're now ready to hack with UTop and follow through this Tour of OCaml. If you hit any issue with the toplevel, don't hesitate ???
```
```suggestion
You're now ready to hack with UTop and follow along through this Tour of OCaml. If you hit any issue with the toplevel, don't hesitate ???
```
Minor suggestion, feel free to ignore. I added "along" in here because I read it as "follow through" as to do something you've promised. 

..."don't hesitate to contact us" ?

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -343,7 +345,7 @@ Lists may be the most common datatype in OCaml. They are ordered collections of
 
 The example above read the following way:
-1. The empty list, prononced “nil”
+1. The empty list, nil
```
```suggestion
1. The empty list (called *nil*)
```

cuihtlauac replied:
Calling the empty list “nil” has already been explained earlier.

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -643,14 +658,14 @@ val id_42_res : int -> (int, string) result = <fun>
 
 It is syntactically impossible to create an unintialised or null reference. The `r`
-reference is initialised with the integer zero. Accessing a reference's content
+reference is initialized with the integer zero. Accessing a reference's content
```
Didn't we decide to keep the UK spelling?

cuihtlauac replied:
Yes we should stick we UK spelling. If it isn't, it's a mistake

christinerose commented on data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md:
```diff
@@ -756,14 +771,14 @@ Module also allows efficient separated compilation, this is illustrated in the n
 ## Sum-Up and Where Next?
 
-OCaml has many features, and these may be its Big Five ones:
+OCaml has many features, and this may be the most important one:
```
This gives a singluar intro "the most important one" but then goes on to list several.

cuihtlauac replied:
Mistake, sorry

christinerose replied:
No worries! 🙂

christinerose committed:
minor grammar fixes

cuihtlauac committed:
data/tutorials/getting-started/gs_01_a_tour_of_ocaml.md updated from https://stackedit.io/

cuihtlauac committed:
Apply suggestions from code review

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

christinerose committed:
UK spelling correction

Commit:
Spelling

christinerose committed:
Add Homebrew link

christinerose committed:
Update gs_04_arm_fix.md

christinerose committed:
punctuation

cuihtlauac commented on data/tutorials/getting-started/gs_06_editor_setup:
TODO: Add vim & Emacs config text

Commit:
Add tutorial on operators

Commit:
Improve operator text

Commit:
Use ChatGPT

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+
+## Using Binary Operators
+
+In OCaml almost all binary operators are regular functions. The function underlying behind an operator is referred by surrounding the operator symbol with parenthesis. Here are the addition, string concatenation and equality functions:
```
```suggestion
In OCaml, almost all binary operators are regular functions. The function underlying behind an operator is referred by surrounding the operator symbol with parentheses. Here are the addition, string concatenation, and equality functions:
```

I'm not clear what you mean by "underlying behind." It's usually one or the other. Is the function behind and operator or underlying it... or maybe even beneath it? Once I understand the meaning, I can revise the sentence.

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
````diff
@@ -0,0 +1,196 @@
+- : int -> int -> int = <fun>
+```
+
+Using operators as functions is convenient when combined with partial application. For instance, here is how to get the values which are greater or equal to 10 in a list of integers, using the function `List.filter` applied to the function `( <= ) 10`:
````
```suggestion
Using operators as functions is convenient when combined with partial application. For instance, use the function `List.filter` applied to the function `( <= ) 10` to get the values greater than or equal to 10 in a list of integers:
```

Do you like this sentence construction better or worse?

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+The first two lines are only informative.
+
+1. The first shows the `List.filter` type, which is a function taking two parameters. The first parameter is a function; the second is a list.
+2. The second is the partial application of `List.filter` to `( <= ) 10`, which is a function returning `true` if applied to a number which is greater or equal than 10.
```
```suggestion
2. The second is the partial application of `List.filter` to `( <= ) 10`, which is a function returning `true` if applied to a number which is greater than or equal to 10.
```

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+
+# let rec transpose = function
+   | [] | [] :: _ -> []
+   | rows -> List.(map hd rows :: transpose (map tl rows))
```
```suggestion
   | rows -> List.(map hd rows :: transpose (map tl rows));;
```

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+
+## Unary Operators
+
+Unary operators are also called prefix operators. In some contexts, it can make sense to shorten the name of a function into a symbol. This is often used as a way to shorten the name of a function which performs some sort of conversion over its argument.
```
```suggestion
Unary operators are also called prefix operators. In some contexts, it can make sense to shorten a function's name into a symbol. This is often used to shorten the name of a function that performs some sort of conversion over its argument.
```

This last sentence it's clear to me. Does shortening the name perform some sort or conversion or does the name of the function itself?

cuihtlauac replied:
It is the function which does a conversion of some sort

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+
+## Allowed Operators
+
+OCaml has a subtle syntax, not everything is allowed as an operator symbol. An operator symbol is an identifier with a spacial syntax, it must have the following structure:
```
```suggestion
OCaml has a subtle syntax, but not everything is allowed as an operator symbol. An operator symbol is an identifier with a spacial syntax, and it must have the following structure:
```

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+
+## Operator Associativity and Precedence
+
+Let's illustrate what operator associativity is using an example. Consider this function, it concatenates its string arguments, surrounded by `|` characters and separated by a `_` character.
```
```suggestion
Let's illustrate operator associativity with an example. Consider the following function. It concatenates its string arguments, surrounded by `|` characters and separated by a `_` character.
```

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
````diff
@@ -0,0 +1,196 @@
+Operator _precedence_ rules how expression combining several operators without parentheses are interpreted. For instance, using the same operators, here is how expressions using both are evaluated.
+```ocaml
+# "foo" &^ "bar" @^ "bus";;
+- : string = "(foo, (bar, bus))"
````
```suggestion
- : string = "|foo_|bar_bus||"
```

I get this

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+- : string = "(foo, (bar, bus))"
+
+# "foo" @^ "bar" &^ "bus";;
+- : string = "((foo, bar), bus)"
```
```suggestion
- : string = "||foo_bar|_bus|"
```

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+1. Expression `"foo" @^ "bar" @^ "bus"` is evaluated as if it was `"foo" @^ ("bar" @^ "bus")`. Parentheses are added at the right, therefore `@^` _associates to the right_
+1. Expression `"foo" &^ "bar" &^ "bus"` is evaluated as if if was `"(foo" @^ "bar") @^ "bus"`. Parentheses are added at the left, therefore `@^` _associates to the left_
+
+Operator _precedence_ rules how expression combining several operators without parentheses are interpreted. For instance, using the same operators, here is how expressions using both are evaluated.
```
```suggestion
Operator _precedence_ rules how expressions that combine several operators without parentheses are interpreted. For instance, expressions using both operators evaluate like this:
```

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
````diff
@@ -0,0 +1,196 @@
+- : string = "((foo, bar), bus)"
+```
+
+In both cases, values are passed to `@^` before `&^`. Therefore, it is said that `@^` has higher _precedence_ over `&^`. Rules for operator priorities are detailed in the [Expressions](https://v2.ocaml.org/manual/expr.html#ss%3Aprecedence-and-associativity) section of the OCaml Manual. They can be summarized the following way. The first character of an operator dictates its associativity and prio ...
````
```suggestion
In both cases, values are passed to `@^` before `&^`. Therefore, it is said that `@^` has higher _precedence_ over `&^`. Rules for operator priorities are detailed in the [Expressions](https://v2.ocaml.org/manual/expr.html#ss%3Aprecedence-and-associativity) section of the OCaml Manual. In short, an operator's first character dictates its associativity and priority. Here are the groups operators of first character. Each group has the same associativity and precedence, and groups are sorted in increasing precedence.
```

I'm stuck on this sentence: "Here are the groups operators of first character. "
Here are the groups' operators' first character.
Here are the group's operators' first character. 
Here are the groups, showing the operators' first character.
... or something else?

If you can help clarify the meaning, I will revise the sentence. Thank you!

christinerose replied:
This one sentence still needs some clarity.

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+## Binding Operators
+
+OCaml allows creation of custom `let` operator. This is often used on functions such as `Option.bind`
+ or `Fun.flip List.concat_map`. In the following example, we use that mechanism to write a function which produces a list of Friday the 13th month between two years.
```
```suggestion
OCaml allows creation of custom `let` operators. This is often used on functions such as `Option.bind`
 or `Fun.flip List.concat_map`. In the following example, we use that mechanism to write a function that produces a list of months containing Friday the 13th over two years.
```

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+# let day_of_week y m d =
+    let t = [| 0; 3; 2; 5; 0; 3; 5; 1; 4; 6; 2; 4 |] in
+    let y = y - if m < 3 then 1 else 0 in
+    (y + y / 4 - y / 100 + y / 400 + t.(m - 1) + d) mod 7
```
```suggestion
    (y + y / 4 - y / 100 + y / 400 + t.(m - 1) + d) mod 7;;
```

christinerose commented on data/tutorials/getting-started/gs_07_operators.md:
```diff
@@ -0,0 +1,196 @@
+# let friday13 lo hi =
+    let* year = range lo hi in
+    let* month = range 1 12 in
+    if day_of_week year month 13 = 5 then [(13, month, year)] else []
```
```suggestion
    if day_of_week year month 13 = 5 then [(13, month, year)] else [];;
```

Commit:
Fix typos

christinerose reviewed (commented):
Caught some missing `;;`, tightened some syntax, and have a few questions about clarity.

christinerose committed:
minor grammar/typos

cuihtlauac committed:
Apply suggestions from code review

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

Commit:
Add sentence

sabine commented on data/tutorials/getting-started/gs_04_arm_fix.md:
```diff
@@ -0,0 +1,69 @@
+---
+id: arm64-fix
+title: Resolve Homebrew Install Errors on ARM64
```
```suggestion
title: Resolve Homebrew Install Errors on ARM64 M1
```

sabine commented on data/tutorials/getting-started/gs_04_arm_fix.md:
```diff
@@ -0,0 +1,69 @@
+  This page will walk you through the workaround for ARM64 processors on newer Macs.
+category: "Troubleshooting"
+---
+
```
```suggestion

# Resolve Homebrew Install Errors on ARM64 M1

```

christinerose committed:
added title

christinerose committed:
add title

Co-authored-by: sabine <sabine@users.noreply.github.com>

christinerose committed:
Update data/tutorials/getting-started/gs_04_arm_fix.md

Co-authored-by: sabine <sabine@users.noreply.github.com>

sabine commented on data/tutorials/getting-started/gs_05_opam_switch.md:
```diff
@@ -0,0 +1,83 @@
+---
+id: opam-switch
+title: Introduction to opam Switch
```
```suggestion
id: opam-switch-introduction
title: Introduction to opam Switches
```

sabine commented on data/tutorials/getting-started/gs_05_opam_switch.md:
```diff
@@ -0,0 +1,83 @@
+  This page will give you a brief introduction to opam switches, what they're used for, and how to creat them.
+category: "Getting Started"
+---
+
```
```suggestion

# Introduction to opam Switches

```

christinerose committed:
Update data/tutorials/getting-started/gs_05_opam_switch.md

Co-authored-by: sabine <sabine@users.noreply.github.com>

christinerose committed:
Update data/tutorials/getting-started/gs_05_opam_switch.md

Co-authored-by: sabine <sabine@users.noreply.github.com>

christinerose committed:
Update gs_05_opam_switch.md

sabine committed:
editing

Commit:
Add links

christinerose committed:
final proofread

christinerose committed:
final proofread

christinerose committed:
edit sentence

christinerose committed:
final proofreading, take 2

christinerose committed:
spell check

christinerose committed:
final proofreading

christinerose committed:
final proofread

christinerose committed:
final proofread

christinerose committed:
final proofread

christinerose committed:
final proofread

christinerose committed:
updated conclusion subheader

Commit:
Change custom let example

Commit:
Change category operators

sabine committed:
cleanups for merging

sabine committed:
rebase mistake fix

sabine committed:
rename, final tweaks
